wallet/walletrpc.c

```diff
@@ -1066,3 +1067,113 @@ static const struct json_command sendpsbt_command = {
 };
 
 AUTODATA(json_command, &sendpsbt_command);
+
+static struct command_result *
+json_signmessagewithkey(struct command *cmd, const char *buffer,
+			const jsmntok_t *obj UNNEEDED, const jsmntok_t *params)
+{
+	/* decoding the address */
+	const u8 *scriptpubkey;
+	const char *message;
+
+	/* from wallet BIP32 */
+	struct pubkey pubkey;
+
+	if (!param(
+		cmd, buffer, params,
+                p_req("message", param_string, &message),
+		p_req("address", param_bitcoin_address, &scriptpubkey),
+                NULL))
+		return command_param_failed();
+
+	const size_t script_len = tal_bytelen(scriptpubkey);
+
+	/* FIXME: we already had the address from the input */
+	char *addr;
+	addr = encode_scriptpubkey_to_addr(tmpctx, chainparams, scriptpubkey);
+
+	if (!is_p2wpkh(scriptpubkey, script_len, NULL)) {
+		/* FIXME add support for BIP 322 */
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				    "Address is not p2wpkh and "
+				    "it is not supported for signing");
+	}
+
+	u32 keyidx = UINT32_MAX;
+
+	/* loop over all generated keys, find a matching key */
+	/* FIXME: alternatively, can we use the our_addresses hash table?
+	 *struct script_with_len *key;
+	 *struct wallet_address *addr =
+	 *    wallet_address_htable_get(cmd->ld->wallet->our_addresses, key);
+	 */
+	struct issued_address_type *listaddrtypes =
+	    wallet_list_addresses(tmpctx, cmd->ld->wallet, 1, NULL);
+	for (size_t i = 0; i < tal_count(listaddrtypes); i++) {
+		bip32_pubkey(cmd->ld, &pubkey, listaddrtypes[i].keyidx);
+		char *out_p2wpkh = "";
+		if (listaddrtypes[i].addrtype == ADDR_BECH32 ||
+		    listaddrtypes[i].addrtype == ADDR_ALL) {
+			u8 *redeemscript_p2wpkh;
+			out_p2wpkh = encode_pubkey_to_addr(
+			    cmd, &pubkey, ADDR_BECH32, &redeemscript_p2wpkh);
+			if (!out_p2wpkh) {
+				abort();
+			}
+		}
+		if (streq(addr, out_p2wpkh)) {
+			keyidx = listaddrtypes[i].keyidx;
+			break;
+		}
+	}
+
+	if (keyidx == UINT32_MAX) {
+		return command_fail(
+		    cmd, JSONRPC2_INVALID_PARAMS,
+		    "Address is not found in the wallet's database");
+	}
+
+	/* wire to hsmd a sign request */
+	u8 *msg = towire_hsmd_sign_message_with_key(
+	    cmd, tal_dup_arr(tmpctx, u8, (u8 *)message, strlen(message), 0),
+	    keyidx);
+	if (!wire_sync_write(cmd->ld->hsm_fd, take(msg))) {
+		fatal("Could not write sign_with_key to HSM: %s",
+		      strerror(errno));
+	}
+
+	/* read form hsmd a sign reply */
+	msg = wire_sync_read(cmd, cmd->ld->hsm_fd);
+
+	int recid;
+	u8 sig[65];
+	secp256k1_ecdsa_recoverable_signature rsig;
+
+	if (!fromwire_hsmd_sign_message_with_key_reply(msg, &rsig)) {
```

Comment:
We should use `hsm_is_capable`, and add the new msg type to capabilities.  This means that an older VLS will not crash: we can tell the user it's not supported.

Reply:
Very important point indeed!

Reply:
[Fixed](https://github.com/ElementsProject/lightning/compare/cbc9205bdc39a0285fd055fa3082466c31941b54..0fcdb2ac5796bd62bd0d0b17f4f7b8696a27a9aa)